src/ATen/native/xpu/sycl/UpSampleLinear1dKernels.cpp

```diff
@@ -0,0 +1,222 @@
+
+    const int batchsize = idata_.size(0);
+    const int channels = idata_.size(1);
+    const int width1 = idata_.size(2);
+    const int width2 = odata_.size(2);
+    PackedTensorAccessor64<scalar_t, 3> odata_res = odata_;
+
+    if (index < n_) {
+      const int w2 = index % width2;
+      // special case: just copy
+      if (width1 == width2) {
+        const int w1 = w2;
+        for (int n = 0; n < batchsize; n++) {
+          for (int c = 0; c < channels; ++c) {
+            const scalar_t val = idata_[n][c][w1];
+            odata_res[n][c][w2] = val;
+          }
+        }
+        return;
+      }
+
+      const accscalar_t w1r = area_pixel_compute_source_index<accscalar_t>(
+          rwidth_, w2, align_corners_, /*cubic=*/false);
+      const int w1 = w1r;
+      const int w1p = (w1 < width1 - 1) ? 1 : 0;
+      const accscalar_t w1lambda = w1r - w1;
+      const accscalar_t w0lambda = static_cast<accscalar_t>(1) - w1lambda;
+
+      for (int n = 0; n < batchsize; n++) {
+        for (int c = 0; c < channels; ++c) {
+          const accscalar_t val =
+              w0lambda * idata_[n][c][w1] + w1lambda * idata_[n][c][w1 + w1p];
+          odata_res[n][c][w2] = static_cast<scalar_t>(val);
+        }
+      }
+    }
+  }
+  UpsampleLinear1dKernelFunctor(
+      const int n,
+      const accscalar_t rwidth,
+      const bool align_corners,
+      const PackedTensorAccessor64<const scalar_t, 3> idata,
+      PackedTensorAccessor64<scalar_t, 3> odata)
+      : n_(n),
+        rwidth_(rwidth),
+        align_corners_(align_corners),
+        idata_(idata),
+        odata_(odata) {}
+
+ private:
+  const int n_;
+  const accscalar_t rwidth_;
+  const bool align_corners_;
+  const PackedTensorAccessor64<const scalar_t, 3> idata_;
+  PackedTensorAccessor64<scalar_t, 3> odata_;
+};
+
+void upsample_linear1d_kernel(
+    const Tensor& input,
+    IntArrayRef output_size,
+    bool align_corners,
+    std::optional<double> scales,
+    Tensor& output) {
+  int output_width = output_size[0];
+  output.zero_();
+  int input_width = input.size(2);
+
+  AT_ASSERT(input_width > 0 && output_width > 0);
+
+  AT_DISPATCH_FLOATING_TYPES_AND2(
+      at::ScalarType::Half,
+      at::ScalarType::BFloat16,
+      input.scalar_type(),
+      "upsample_linear1d_kernel",
+      [&] {
+        auto idata = input.packed_accessor64<const scalar_t, 3>();
+        auto odata = output.packed_accessor64<scalar_t, 3>();
+
+        using accscalar_t = at::acc_type_device<scalar_t, kXPU>;
+        const accscalar_t rwidth = area_pixel_compute_scale<accscalar_t>(
+            input_width, output_width, align_corners, scales);
+        const int num_kernels = output_width;
+        using KernelClass =
+            UpsampleLinear1dKernelFunctor<scalar_t, accscalar_t>;
+        int64_t local_range = syclMaxWorkGroupSize<KernelClass>();
+        KernelClass kfn(num_kernels, rwidth, align_corners, idata, odata);
```

Comment:
Suggest using `syclMaxWorkGroupSize(kfn)` not forward-declaring `KernelClass`, if ctor of kernel functor class doesn't depend on work group size.